### PR TITLE
Optimize performance and memory usage in dynamics computations

### PR DESCRIPTION
**Pull request by flferretti**
Enhance efficiency by normalizing quaternions, reducing memory footprint, and speeding up various dynamics computations. Replace `jnp.float32` with `float` for consistency. Avoid unnecessary calls to `.names_to_idxs` where possible.
> [!Note]
> See profiler traces at https://gist.github.com/flferretti/a20c7cd171c25dd784eef361246bd8cd

> [!Tip]
> - `js.model.generalized_free_floating_jacobian_derivative` used `jax.vmap` over `js.link.jacobian_derivative`, but a lot of quantities were unnecessarily computed multiple times. For this reason, I moved the major computation to `js.model` and left only the link extraction part in `js.link`. Some adjustments regarding the output velocity representation were necessary in order to avoid to call `js.link.velocity` inside the `vmap`ped function `_compute_row`

| Function                                                         | New Compile Time | Original Compile Time | New Run Time | Original Run Time |
|------------------------------------------------------------------|------------------|-----------------------|--------------|-------------------|
| `js.model.generalized_free_floating_jacobian_derivative`         | 643 ms           | 748 ms                | **479 μs**      | 6.07 ms           |
| `js.model.free_floating_mass_matrix`                             | 423 ms           | 456 ms                | **470 μs**       | 502 μs            |
| `js.model.forward_kinematics`                                    | 246 ms           | 255 ms                | **406 μs**       | 445 μs            |
| `js.model.generalized_free_floating_jacobian`                    | 311 ms           | 368 ms                | **455 μs**       | 473 μs            |
| `js.references.JaxSimModelReferences.build`                      | 98.3 ms          | 139 ms                | **558 μs**       | 498 μs            |
| `js.contact.collidable_point_kinematics`                         | 326 ms           | 332 ms                | **420 μs**       | 432 μs            |
| `data.base_orientation`                         | 14.3 ms          | 19.1 ms               | **10.2 μs**      | 14.6 μs           |
| `model.kin_dyn_parameters.joint_transforms_and_motion_subspaces` | 236 ms           | 241 ms                | **53.3 μs**      | 60.1 μs           |
| `references.link_forces`                                         | 12.7 ms          | 38.2 ms               | **20.9 μs**      | 22.1 μs           |

----
📚 Documentation preview 📚: https://jaxsim--280.org.readthedocs.build//280/

### CODE REVIEW

**Review by diegoferigo — COMMENTED** [review]
Leaving a preliminary comment before going more in detail with a review.

Do you have any benchmark supporting the need to switch from names to indices? Using indices makes the code much more fragile and, if the indexing logic changes, there will be countless of places to update. The name-to-index helpers provide a centralized location to store this logic.

A very descriptive example of what I just said is the joint name to index map. ~There is a tricky +- 1 index, that at first sight this PR lacks.~ I'm not that supportive of gaining minor performance benefits if the price to pay is less maintainable and more error prone code. However, I'll be waiting for some benchmark, we can assess the situation more clearly with data.

Edit: I double checked the indexing returned by `jaxsim.api.joint.names_to_indices` and it actually starts from 0. It's the low-level indexing used by RBDAs that start from 1.

**Comment by flferretti** [comment]
I'm doing the benchmarks using Perfetto Profiler (see https://jax.readthedocs.io/en/latest/profiling.html). I'll post them ASAP

**Comment by diegoferigo** [comment]
> > Leaving a preliminary comment before going more in detail with a review.
> > Do you have any benchmark supporting the need to switch from names to indices? Using indices makes the code much more fragile and, if the indexing logic changes, there will be countless of places to update. The name-to-index helpers provide a centralized location to store this logic.
> > A very descriptive example of what I just said is the joint name to index map. There is a tricky +- 1 index, that at first sight this PR lacks. I'm not that supportive of gaining minor performance benefits if the price to pay is less maintainable and more error prone code. However, I'll be waiting for some benchmark, we can assess the situation more clearly with data.
> 
> I haven't removed the usage of `*_names`, but simply avoided to call `js.*.names_to_idxs` when `*_names` is not passed to the call. In that case, we explicitly need all of the `*` (joint/link/frame), so we can just use `model.number_of_*`

Yes you introduced the knowledge that you can use `arange` in that case. But as said, this change spreads around this piece of knowledge that might change in the future (what if we start supporting free-floating joints to unify the base-related section of RBDAs and the corresponding base joint get a zero index?). If there are no relevant performance consideration in place, why making the code more hard to read and error prone for future refactorings? (this is also a general comment, not necessarily related to just this change).

**Comment by flferretti** [comment]
> But as said, this change spreads around this piece of knowledge that might change in the future (what if we start supporting free-floating joints to unify the base-related section of RBDAs and the corresponding base joint get a zero index?).

I believe that the most intuitive behavior for the final user is that 
```python
js.link.names_to_indexes(joint_names=model.link_names(), **kwargs)
```
and 
```python
jnp.arange(model.number_of_links())
```
 should return the same result. If there will be any need to change this logic in the future, we will propagate the change over the whole library

**Comment by diegoferigo** [comment]
> > But as said, this change spreads around this piece of knowledge that might change in the future (what if we start supporting free-floating joints to unify the base-related section of RBDAs and the corresponding base joint get a zero index?).
> 
> I believe that the most intuitive behavior for the final user is that
> 
> ```python
> js.link.names_to_indexes(joint_names=model.link_names(), **kwargs)
> ```
> 
> and
> 
> ```python
> jnp.arange(model.number_of_links())
> ```
> 
> should return the same result. If there will be any need to change this logic in the future, we will propagate the change over the whole library

My point is not on the user side since users are not supposed to know anything about indices (there are the names to index helpers for this reason). My point was about the developers and the logic within JaxSim functions.

I'm not strongly against this change, my whole point is that in case of changes on the indices used internally, there will be the need to change multiple locations. I checked the tests, and we already have checks that ensure that the current indexing does not get broken:

https://github.com/ami-iit/jaxsim/blob/bd590609ab72f639e733446270e2b26a974904a6/tests/test_api_joint.py#L22-L24

https://github.com/ami-iit/jaxsim/blob/main/tests/test_api_link.py#L27-L29

Can you please add a comment to both files that links to our discussion in this PR in case that check starts to fail?

File: src/jaxsim/math/adjoint.py [hunk 1]
```diff
@@ -137,11 +137,12 @@ def inverse(adjoint: jtp.Matrix) -> jtp.Matrix:
...
+        A_R_B = A_X_B[0:3, 0:3]
 
-        return Adjoint.from_rotation_and_translation(
-            rotation=A_R_B, translation=A_o_B, inverse=True
+        return jnp.vstack(
+            [
+                jnp.block([A_R_B.T, -A_R_B.T @ A_X_B[0:3, 3:6] @ A_R_B.T]),
+                jnp.block([jnp.zeros(shape=(3, 3)), A_R_B.T]),
+            ]
         )
```
Comment (diegoferigo) [inline src/jaxsim/math/adjoint.py:148]: I just realized that all these methods are correct only if they operate on transforms for 6D velocites, not those for 6D forces. Let's keep this in mind. We might also want to add an additional argument to turn on logic for 6D forces. I think that the term _adjoint_ only corresponds to the tangent of the original space ($\text{SE}(3) in this case), and it might mean that the considered adjoint is the trasform for 6D velocities. Not sure if the real terminology for the trasform for 6D forces is _dual_ adjoint.  cc @traversaro

**Comment by flferretti** [comment]
> Can you please add a comment to both files that links to our discussion in this PR in case that check starts to fail?

Done in https://github.com/ami-iit/jaxsim/pull/280/commits/85ab07ca445c955c1f4c3dcdaa7d6ad75c396922. Thanks!

**Comment by flferretti** [comment]
Traces available at https://gist.github.com/flferretti/a20c7cd171c25dd784eef361246bd8cd.

In order to view them, please run:

```bash
git clone https://gist.github.com/a20c7cd171c25dd784eef361246bd8cd.git
cd a20c7cd171c25dd784eef361246bd8cd
python -m http.server
```

C.C. @xela-95 @diegoferigo

File: src/jaxsim/math/adjoint.py (hunk 1 above)
Comment (traversaro) [reply src/jaxsim/math/adjoint.py:148]: Murray Lee Sastry seems just to write of "adjoint transpose". Indeed Park calls it "dual adjoint" in https://mae.eng.uci.edu/~bobrow/papers_files/DynamicsParkBobrow.pdf , but if you want to avoid confusion in the reader you can also use more user friendly terms like "velocity transform" and "force transform" or something similar.

**Comment by diegoferigo** [comment]
> Traces available at https://gist.github.com/flferretti/a20c7cd171c25dd784eef361246bd8cd.
> 
> In order to view them, please run:
> 
> ```shell
> git clone https://gist.github.com/a20c7cd171c25dd784eef361246bd8cd.git
> cd a20c7cd171c25dd784eef361246bd8cd
> python -m http.server
> ```
> 
> C.C. @xela-95 @diegoferigo

Can you please summarize them, if you have any specific improvement in mind?

**Comment by flferretti** [comment]
> Can you please summarize them, if you have any specific improvement in mind?

I made the gist just for reference and in case you were interested into the specific improvements. For a general benchmark, you can take a look at the PR description

File: src/jaxsim/api/data.py [hunk 2]
```diff
@@ -374,11 +378,7 @@ def base_orientation(self, dcm: jtp.BoolLike = False) -> jtp.Vector | jtp.Matrix
         # we introduce a Baumgarte stabilization to let the quaternion converge to
         # a unit quaternion. In this case, it is not guaranteed that the quaternion
         # stored in the state is a unit quaternion.
-        W_Q_B = jax.lax.select(
-            pred=jnp.allclose(jnp.linalg.norm(W_Q_B), 1.0, atol=1e-6, rtol=0.0),
-            on_true=W_Q_B,
-            on_false=W_Q_B / jnp.linalg.norm(W_Q_B),
-        )
+        W_Q_B = W_Q_B / jnp.linalg.norm(W_Q_B)
```
Comment (diegoferigo) [inline src/jaxsim/api/data.py:381]: For AD purpose, having a division with a square root is suboptimal. I'd try to apply normalization only if needed:

```suggestion
        W_Q_B = jnp.where(
            W_Q_B.dot(W_Q_B) == 1.0, W_Q_B, W_Q_B / jnp.linalg.norm(W_Q_B)
        )
```

**Review by diegoferigo — APPROVED** [review]
Great, thanks @flferretti! The improvement of $\dot{J}$ was much needed.

File: src/jaxsim/api/data.py (hunk 2 above)
Comment (flferretti) [reply src/jaxsim/api/data.py]: Thanks! Addressed in https://github.com/ami-iit/jaxsim/pull/280/commits/03edbf2a6a1c6ca63711312b04e349cb02312392

Comment (diegoferigo) [reply src/jaxsim/api/data.py]: Should be use allclose or == 1? Strict comparison with floats are tricky. I think we can support quaternions with a norm within the default tolerances of allclose.

Comment (flferretti) [reply src/jaxsim/api/data.py]: Addressed with rebase

**Review by xela-95 — APPROVED** [review]
Thanks @flferretti for the performance improvements!! I checked your changes and they look good to me!
